notarization_poller/src/notarization_poller/worker.py

```diff
@@ -0,0 +1,152 @@
+                        options={
+                            "credentials": {"accessToken": self.config["taskcluster_access_token"], "clientId": self.config["taskcluster_client_id"]},
+                            "rootUrl": self.config["taskcluster_root_url"],
+                        },
+                        session=session,
+                    )
+                    new_tasks = await self._run_cancellable(claim_work(self.config, queue, num_tasks=num_tasks_to_claim)) or {}
+                self.last_claim_work = arrow.utcnow()
+                for claim_task in new_tasks.get("tasks", []):
+                    new_task = Task(self.config, claim_task)
+                    new_task.start()
+                    self.running_tasks.append(new_task)
+            await self.prune_running_tasks()
+            sleep_time = self.last_claim_work.timestamp + self.config["claim_work_interval"] - arrow.utcnow().timestamp
+            sleep_time > 0 and await self._run_cancellable(sleep(sleep_time))
+        self.running_tasks and await asyncio.wait([task.main_fut for task in self.running_tasks if task.main_fut])
+
+    async def prune_running_tasks(self):
+        """Prune any complete tasks from ``self.running_tasks``."""
+        for task in self.running_tasks:
+            if task.complete:
+                self.running_tasks.remove(task)
+
+    async def _run_cancellable(self, coroutine: typing.Awaitable):
+        if not self.is_cancelled:
+            self.future = asyncio.ensure_future(coroutine)
+            result = await self.future
+            self.future = None
+            return result
+
+    async def cancel(self, status=STATUSES["worker-shutdown"]):
+        """Cancel any running tasks."""
+        self.is_cancelled = True
+        self.future and self.future.cancel()
+        try:
+            for task in self.running_tasks:
+                task.task_fut and task.task_fut.cancel()
+            await asyncio.wait([task.main_fut for task in self.running_tasks if task.main_fut])
+        except (asyncio.CancelledError, ValueError):
+            pass
+
+
+# main {{{1
+def main(event_loop=None):
+    """Notarization poller entry point: get everything set up, then enter the main loop.
+
+    Args:
+        event_loop (asyncio.BaseEventLoop, optional): the event loop to use.
+            If None, use ``asyncio.get_event_loop()``. Defaults to None.
+
+    """
+    event_loop = event_loop or asyncio.get_event_loop()
+    done = False
+    config = get_config_from_cmdln(sys.argv[1:])
+    update_logging_config(config)
+
+    log.info("Notarization poller starting up at {} UTC".format(arrow.utcnow().format()))
+    log.info("Worker FQDN: {}".format(socket.getfqdn()))
+    rm(config["work_dir"])
+    makedirs(config["work_dir"])
+    running_tasks = RunTasks(config)
+
+    async def _handle_sigterm():
+        log.info("SIGTERM received; shutting down")
+        nonlocal done
+        done = True
+        await running_tasks.cancel()
```

Comment:
Since `add_signal_handler` does handle futures, I think it would be better if `_handle_sigterm` were sync and handled anything that needed to be async itself.

That said, it looks like it async so that `running_tasks.cancel` can wait on the task futures. Since nothing is waiting on the result of that, I don't think that await actually has any effect. Thus, I think this function and `running_tasks.cancel` can both be sync.

https://github.com/mozilla-releng/scriptworker-scripts/commit/fae37f8e48fae5e9beebe51c7b017fc5e6383324#diff-5fb318e79005ec63cfed70064fa1e861R71 is where I handled waiting on all of those tasks to finish in my sketch

Reply:
I mainly copied over [these coroutines](https://github.com/mozilla-releng/scriptworker/blob/master/src/scriptworker/worker.py#L236). I'm almost leaning towards tearing out this code, since the worst that can happen with `notarization_poller` dying is we wait a bit longer on tasks that just poll and wait. Do you have a preference between tearing these out and leaving them as-is?

Reply:
RIght. My main concerns are: I have done a significant amount of testing with the code as is. Making non-trivial changes has the likelihood of introducing new errors, just as I'm about to roll out.